### PR TITLE
chore: bump ot version to 0.31.0

### PR DESCRIPTION
**Pull request by pmalek-sumo**
This uses core fork from [main-rebase-tip-0.31.0](https://github.com/SumoLogic/opentelemetry-collector/tree/main-rebase-tip-0.31.0) branch (specifically `v0.31.0-sumo-1` tag).

List of breaking changes this needed work on:
* https://github.com/open-telemetry/opentelemetry-collector/releases/tag/v0.30.0
* https://github.com/open-telemetry/opentelemetry-collector/releases/tag/v0.31.0

### CODE REVIEW

File: pkg/processor/sourceprocessor/source_processor.go [hunk 1]
```diff
@@ -247,7 +247,7 @@ func (stp *sourceTraceProcessor) ConsumeTraces(ctx context.Context, td pdata.Tra
 		}
 
 		if stp.isFilteredOut(atts) {
-			rs.InstrumentationLibrarySpans().Resize(0)
+			rs.InstrumentationLibrarySpans().RemoveIf(func(pdata.InstrumentationLibrarySpans) bool { return true })
```
Comment (sumo-drosiek) [inline pkg/processor/sourceprocessor/source_processor.go:250]: does remove if return how many spans have been removed?

File: pkg/processor/sumologicsyslogprocessor/processor_test.go [hunk 2]
```diff
@@ -38,16 +38,15 @@ func TestProcessLogs(t *testing.T) {
...
+	lgs := rls.InstrumentationLibraryLogs().AppendEmpty()
 
 	for _, line := range lines {
-		lr := pdata.NewLogRecord()
+		lr := lgs.Logs().AppendEmpty()
 		lr.Body().SetStringVal(line)
-		lgs.Append(lr)
 	}
-	lgs.At(1).Attributes().InsertString("facility_name", "pre filled facility")
+	lgs.Logs().At(1).Attributes().InsertString("facility_name", "pre filled facility")
```
Comment (sumo-drosiek) [inline pkg/processor/sumologicsyslogprocessor/processor_test.go:49]: `lgs` can be misleading now

File: pkg/processor/sourceprocessor/source_processor.go (hunk 1 above)
Comment (pmalek-sumo) [reply pkg/processor/sourceprocessor/source_processor.go:250]: Nope, 

```go
func (es InstrumentationLibrarySpansSlice) RemoveIf(f func(InstrumentationLibrarySpans) bool) {
```

File: pkg/processor/sumologicsyslogprocessor/processor_test.go (hunk 2 above)
Comment (pmalek-sumo) [reply pkg/processor/sumologicsyslogprocessor/processor_test.go:49]: Granted that it's `InstrumentationLibraryLogs` I could call it `ils`, WDYT?

Comment (pmalek-sumo) [reply pkg/processor/sumologicsyslogprocessor/processor_test.go]: Done

**Review by sumo-drosiek — APPROVED** [review]

**Pull request merged**